pip/configuration.py

```diff
@@ -4,68 +4,245 @@
+
+
     """
 
-    def __init__(self, isolated):
-        self._configparser = configparser.RawConfigParser()
-        self._config = {}
+    def __init__(self, isolated, load_only=None):
+        super(Configuration, self).__init__()
+
+        if load_only not in ["user", "site-wide", "venv", None]:
+            raise ValueError(
+                "Got invalid value for load_only - should be one of 'user', "
+                "'site-wide', 'venv'"
+            )
         self.isolated = isolated
+        self.load_only = load_only
 
-    def load(self, section):
-        """Loads configuration
+        # The order here determines the override order.
+        self._override_order = ["site-wide", "user", "venv", "environment"]
+        # Because we keep track of where we got the data from
+        self._parsers = {variant: [] for variant in self._override_order}
+        self._config = {variant: {} for variant in self._override_order}
+        self._modified_parsers = []
+
+    def load(self):
+        """Loads configuration from configuration files and environment
         """
-        self._load_config_files(section)
+        self._load_config_files()
         if not self.isolated:
             self._load_environment_vars()
 
     def items(self):
-        """Returns key-value pairs like dict.values() representing the loaded
+        """Returns key-value pairs like dict.items() representing the loaded
         configuration
         """
-        return self._config.items()
+        return self._dictionary.items()
+
+    #
+    # Exposed only for config command. Other commands should not touch this.
+    #
+
+    def get_value(self, key):
+        """Get a value from the configuration.
+        """
+        return self._dictionary[key]
+
+    def set_value(self, key, value):
+        """Modify a value in the configuration.
+        """
+        assert self.load_only is not None, _need_file_err_msg
+
+        file, parser = self._get_parser_to_modify()
+
+        if parser is not None:
+            section, name = _disassemble_key(key)
+
+            # Modify the parser and the configuration
+            if not parser.has_section(section):
+                parser.add_section(section)
+            parser.set(section, name, value)
+
+        self._config[self.load_only][key] = value
+
+        # MARK:: Maybe DRY this.
+        file_parser_tuple = (file, parser)
+        if file_parser_tuple not in self._modified_parsers:
+            self._modified_parsers.append(file_parser_tuple)
+
+    def unset_value(self, key):
+        """Unset a value in the configuration.
+        """
+        assert self.load_only is not None, _need_file_err_msg
+
+        if key not in self._config[self.load_only]:
+            raise KeyError(key)
+
+        file, parser = self._get_parser_to_modify()
+
+        if parser is not None:
+            section, name = _disassemble_key(key)
+
+            # Remove the key in the parser
+            modified_something = (
```

Comment:
This is a bit hard to follow, I generally prefer not to have methods with side effects in a boolean expression because it's easy to miss what's going on. I think this would be more readable as:

```python
modified_something = False
if parser_has_section(section):
    modified_something = parser.remove_option(section, name)
```

I actually missed that this was having any side effect at first and it took me a minute to understand why you wanted a comment here to start with.